bootloader/src/old_machine_bootloader_linux.c

```diff
@@ -0,0 +1,202 @@
+#include <arpa/inet.h>
+#include <ifaddrs.h>
+#include <curl/curl.h>
+#include "old_machine_common_functions.h"
+
+#define BOOTLOADER_SERVER_PORT ":5001"
+#define ISLAND_SERVER_PORT ":5000"
+
+void error(const char *msg) { perror(msg); exit(1); }
+
+struct response {
+  char *ptr;
+  size_t len;
+};
+
+void init_response(struct response *s) {
+  s->len = 0;
+  s->ptr = malloc(s->len+1);
+  if (s->ptr == NULL)
+    error("malloc() failed\n");
+  s->ptr[0] = '\0';
+}
+
+size_t writefunc(void *ptr, size_t size, size_t nmemb, struct response *s)
+{
+  size_t new_len = s->len + size*nmemb;
+  s->ptr = realloc(s->ptr, new_len+1);
+  if (s->ptr == NULL)
+    error("realloc() failed\n");
+
+  memcpy(s->ptr+s->len, ptr, size*nmemb);
+  s->ptr[new_len] = '\0';
+  s->len = new_len;
+
+  return size*nmemb;
+}
+
+char* executeCommand(char* commandLine){
+    FILE *fp;
+    const int maxOutputLength = 2400;
+    char* path = (char *) malloc(maxOutputLength);
+
+    /* Open the command for reading. */
+    fp = popen(commandLine, "r");
+    if (fp == NULL)
+        error("Failed to run command\n" );
+
+    /* Read the output a line at a time - output it. */
+    char* res = fgets(path, maxOutputLength, fp);
+    if (res == NULL)
+        error("ERROR reading commandline\n");
+
+    /* close */
+    pclose(fp);
+
+    return path;
+}
+
+struct response sendRequest(char* server, char* tunnel, char* data){
+    CURL *curl;
+    CURLcode res;
+    struct response s;
+    curl = curl_easy_init();
+    init_response(&s);
+    if(curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, server);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writefunc);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &s);
+        if(tunnel != NULL){
+            curl_easy_setopt(curl, CURLOPT_PROXY, tunnel);
+        }
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+        /* Perform the request, res will get the return code */
+        res = curl_easy_perform(curl);
+        /* Check for errors */
+        if(res != CURLE_OK){
+            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+            s.ptr = "FAILED";
+            s.len = strlen(s.ptr);
+            return s;
+        }
+
+        /* always cleanup */
+        curl_easy_cleanup(curl);
+    }
+    curl_global_cleanup();
+    return s;
+}
+
+char** getIpAddresses(int *addrCount){
+    char** IPs = NULL;
+    int i = 0;
+
+    struct ifaddrs * ifAddrStruct=NULL;
+    struct ifaddrs * ifa=NULL;
+    void * tmpAddrPtr=NULL;
+    getifaddrs(&ifAddrStruct);
+
+    for (ifa = ifAddrStruct; ifa != NULL; ifa = ifa->ifa_next) {
+        if (!ifa->ifa_addr || i >= maxSize) {
+            continue;
+        }
+        if (ifa->ifa_addr->sa_family == AF_INET) { // check it is IP4
+            // is a valid IP4 Address
+            tmpAddrPtr=&((struct sockaddr_in *)ifa->ifa_addr)->sin_addr;
+            i = i+1;
+            IPs = (char**)realloc(IPs, (i+1)*sizeof(*IPs));
```

Comment:
You can save these realloc and malloc by iterating over `ifa` once and counting the number of valid addresses, allocating all the data you need and then iterating again to copy the data.
the reallocs are more expensive than iterating over the list twice,

Reply:
Good to know, but I don't think it matters much in this context.

Reply:
No it doesn't, but it's cleaner code.